lib/validator.js

```diff
@@ -0,0 +1,340 @@
+    var policy = validateAndNormalizeRevPolicy(schema);
+    if (policy) {
+        schema.revisionRetentionPolicy = policy;
+    }
+
+    // XXX: validate attributes
+    return schema;
+};
+
+/**
+ * Validates a get request projection. In case any of the following preconditions fail,
+ * {Error} is thrown.
+ *
+ * - A projection must be a String or Array
+ * - All projection attributes must exist in a table/index schema
+ * @param proj a projection to validate
+ * @param schema original table schema
+ */
+function validateProj(proj, schema) {
+    proj = proj || schema.proj;
+    if (!Array.isArray(proj) && proj.constructor !== String) {
+        throw new Error('Invalid query. Projection of type ' + proj.constructor + ' not supported');
+    }
+    if (proj.constructor === String) {
+        proj = [ proj ];
+    }
+    if (proj.some(function(projElem) {
+        return Object.keys(schema.attributes).indexOf(projElem) < 0;
+    })) {
+        throw new Error('Invalid query. Projection element ' + projElem + ' not in the schema');
+    }
+}
+
+/**
+ * Validates a predicate option of a get query. If any of the following preconditions
+ * fail, {Error} is thrown.
+ *
+ * - If primaryKeyOnly is true, every attribute in the predicate must be 'hash' or 'range' indexed
+ * - Each individual condition operator must be valid
+ * - Non-eq operators allowed only on 'range' indexed columns
+ * - There can't be more than one non-eq predicate for different columns
+ * - There can't be an eq predicate after a non-eq predicate
+ *
+ * @param pred a predicate object from the query
+ * @param schema original table schema
+ * @param primaryKeyOnly indicates if it should be verified that each key is a primary key
+ */
+function validatePredicate(pred, schema, primaryKeyOnly) {
+    var nonEqFound;
+    Object.keys(pred).forEach(function(predKey) {
+        var predObj = pred[predKey];
+        // Check tht a key is in hash/range index
+        if (primaryKeyOnly && (!schema.iKeyMap[predKey] ||
+                (schema.iKeyMap[predKey].type !== 'hash'
+                && schema.iKeyMap[predKey].type !== 'range'))) {
+            throw new Error('Invalid query. attribute ' + predKey
+                + " is not a part of primary key and can't be in condition")
+        }
+        if (predObj === undefined) {
+            throw new Error('Invalid query. attribute ' + predKey + ' is undefined');
+        } else if (predObj.constructor === Object) {
+            var predOp = Object.keys(predObj)[0].toLowerCase();
+            if (validPredKeys.indexOf(predOp) < 0) {
+                throw new Error('Illegal predicate operator for ' + JSON.stringify(predObj));
+            }
+            if (predOp !== 'eq') {
+                if (schema.iKeyMap[predKey].type !== 'range') {
+                    throw new Error('Invalid query. Non-eq conditions allowed only on range columns: ' + predKey);
+                }
+                if (nonEqFound !== undefined) {
```

Comment:
If on L251 the `else if` is promoted to a proper `if`, then this one can be removed, as they do the same thing. Code clean-up FTW